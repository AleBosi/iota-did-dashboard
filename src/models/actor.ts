// /models/actor.ts

export interface Actor {
  id: string; // DID unico (es: "did:iota:evm:0x...")
  name: string;
  role: "admin" | "azienda" | "creator" | "operatore" | "macchinario";
  aziendaId?: string;      // DID dell’azienda di appartenenza
  seed?: string;           // non esportare mai in VC!
  publicKey?: string;      // per interoperabilità/validazione
  createdAt?: string;      // opzionale, ISO date
  updatedAt?: string;      // opzionale

  // AGGIUNGI QUESTO CAMPO:
  vcIds?: string[];        // Lista degli ID delle VC associate a questo attore
}
