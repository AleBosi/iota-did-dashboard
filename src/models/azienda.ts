// /models/azienda.ts
export interface Azienda {
  id: string;         // DID (es: did:iota:evm:...)
  name: string;
  seed: string;       // per login e firma
  legalInfo?: {
    vat?: string;         // partita IVA o VAT
    lei?: string;         // Legal Entity Identifier (EBSI ready)
    address?: string;
    email?: string;
    country?: string;
    [key: string]: any;   // altri campi liberi
  };
  creators: string[];     // array di DID
  operatori: string[];
  macchinari: string[];
  createdAt?: string;     // ISO
  updatedAt?: string;     // ISO
}
