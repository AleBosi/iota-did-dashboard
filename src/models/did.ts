// /models/did.ts
export interface DidDocument {
  id: string; // es: "did:iota:evm:0x123..."
  controller?: string[];
  verificationMethod?: {
    id: string;
    type: string;
    publicKeyMultibase: string;
  }[];
  authentication?: string[];
  service?: {
    id: string;
    type: string;
    serviceEndpoint: string;
  }[];
  // Altri campi secondo specifica, se serve
}
