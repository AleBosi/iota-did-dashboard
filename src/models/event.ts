// /models/event.ts

export interface Event {
  id: string;
  productId: string;         // ID prodotto collegato (ora obbligatorio)
  operatoreId: string;       // ID operatore coinvolto
  macchinarioId: string;     // ID macchinario coinvolto
  type: string;              // es: "Assemblaggio", "Controllo", "Assegnazione"
  description: string;
  date: string;              // ISO
  creatorId: string;         // DID del creator che ha creato lâ€™evento
  done?: boolean;            // true se eseguito
  bomComponent?: string;     // opzionale, componente specifico
  proofId?: string;          // se firmerai evento come VC
  vcIds?: string[];          // VC associate/notarizzate collegate a questo evento
}
