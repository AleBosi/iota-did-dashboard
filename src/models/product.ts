// /models/product.ts
export interface Product {
  id: string;              // ID univoco locale/app
  did: string;             // DID prodotto (es: did:iota:evm:0x...)
  name: string;            // Nome commerciale o etichetta prodotto
  typeId?: string;         // riferimento a ProductType.id
  serial?: string;         // seriale prodotto
  owner?: string;          // DID del proprietario corrente
  bom?: { id: string; name: string }[]; // lista componenti (Bill of Materials)
  credentials?: string[];  // lista di VC associate (ID delle VC)
  description?: string;    // opzionale
}
