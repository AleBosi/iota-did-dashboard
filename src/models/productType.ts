// /models/productType.ts
export interface ProductType {
  id: string;               // es: GTIN, oppure UUID univoco
  name: string;
  description?: string;
}
