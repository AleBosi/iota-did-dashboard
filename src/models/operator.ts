// /models/operator.ts
import { Actor } from "./actor";

export interface Operator extends Actor {
  // aggiungi campi solo se servono SOLO agli operatori
}
