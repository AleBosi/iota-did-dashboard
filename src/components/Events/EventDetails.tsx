import { Event } from "../../models/event";
import { Actor } from "../../models/actor";

interface Props {
  event: Event;
  actors?: Actor[]; // opzionale: se vuoi mostrare nomi/DID reali invece di solo ID
}

export default function EventDetails({ event, actors }: Props) {
  if (!event) return null;

  // Helper per trovare attore con id
  const findActor = (id?: string) => actors?.find(a => a.id === id);

  // Trova dettagli operatore/macchinario
  const operatore = findActor?.(event.operatoreId);
  const macchinario = findActor?.(event.macchinarioId);

  return (
    <div className="border rounded p-4 bg-gray-50 mb-2">
      <div><b>ID evento:</b> {event.id}</div>
      <div><b>Tipo:</b> {event.type}</div>
      <div><b>Descrizione:</b> {event.description}</div>
      <div><b>Prodotto:</b> {event.productId}</div>

      <div>
        <b>Operatore:</b>{" "}
        {operatore ? (
          <>
            {operatore.name} <span className="text-xs text-gray-500">({operatore.did})</span>
          </>
        ) : (
          event.operatoreId
        )}
      </div>

      <div>
        <b>Macchinario:</b>{" "}
        {macchinario ? (
          <>
            {macchinario.name} <span className="text-xs text-gray-500">({macchinario.did})</span>
          </>
        ) : (
          event.macchinarioId
        )}
      </div>

      <div><b>Creatore (DID):</b> {event.creatorId}</div>
      {event.bomComponent && <div><b>Componente BOM:</b> {event.bomComponent}</div>}
      <div><b>Data:</b> {new Date(event.date).toLocaleString()}</div>
      <div>
        <b>Stato:</b>{" "}
        <span className={event.done ? "text-green-700 font-bold" : "text-orange-700"}>
          {event.done ? "Completato" : "Da eseguire"}
        </span>
      </div>
      {event.proofId && <div><b>VC principale (proofId):</b> {event.proofId}</div>}
      {event.vcIds && event.vcIds.length > 0 && (
        <div>
          <b>VC associate:</b> {event.vcIds.join(", ")}
        </div>
      )}
    </div>
  );
}
