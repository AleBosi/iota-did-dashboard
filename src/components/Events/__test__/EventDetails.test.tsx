import { render, screen } from "@testing-library/react";
import EventDetails from "../EventDetails";
import { Event } from "../../../models/event";

const baseEvent: Event = {
  id: "evt-001",
  type: "manutenzione",
  description: "Controllo annuale macchina",
  productId: "prod-01",
  operatoreId: "op-11",
  macchinarioId: "mac-22",
  creatorId: "did:creator-xyz",
  date: "2025-08-05T12:00:00Z",
  done: false
};

describe("EventDetails component", () => {
  it("renderizza tutti i campi base obbligatori", () => {
    render(<EventDetails event={baseEvent} />);
    expect(screen.getByText(/ID evento:/i)).toBeInTheDocument();
    expect(screen.getByText("evt-001")).toBeInTheDocument();

    expect(screen.getByText(/Tipo:/i)).toBeInTheDocument();
    expect(screen.getByText("manutenzione")).toBeInTheDocument();

    expect(screen.getByText(/Descrizione:/i)).toBeInTheDocument();
    expect(screen.getByText("Controllo annuale macchina")).toBeInTheDocument();

    expect(screen.getByText(/Prodotto:/i)).toBeInTheDocument();
    expect(screen.getByText("prod-01")).toBeInTheDocument();

    expect(screen.getByText(/Operatore:/i)).toBeInTheDocument();
    expect(screen.getByText("op-11")).toBeInTheDocument();

    expect(screen.getByText(/Macchinario:/i)).toBeInTheDocument();
    expect(screen.getByText("mac-22")).toBeInTheDocument();

    expect(screen.getByText(/Creatore \(DID\):/i)).toBeInTheDocument();
    expect(screen.getByText("did:creator-xyz")).toBeInTheDocument();

    expect(screen.getByText(/Data:/i)).toBeInTheDocument();
    // La data è ora formattata localmente, quindi regex (match qualsiasi data)
    expect(screen.getByText(/\d+\/\d+\/\d+/)).toBeInTheDocument();

    expect(screen.getByText(/Stato:/i)).toBeInTheDocument();
    expect(screen.getByText(/Da eseguire/i)).toBeInTheDocument();
  });

  it("mostra 'Completato' se done è true", () => {
    render(<EventDetails event={{ ...baseEvent, done: true }} />);
    expect(screen.getByText(/Stato:/i)).toBeInTheDocument();
    expect(screen.getByText(/Completato/i)).toBeInTheDocument();
  });

  it("visualizza campo opzionale bomComponent se presente", () => {
    render(<EventDetails event={{ ...baseEvent, bomComponent: "motore-01" }} />);
    expect(screen.getByText(/Componente BOM:/i)).toBeInTheDocument();
    expect(screen.getByText("motore-01")).toBeInTheDocument();
  });

  it("visualizza campo opzionale proofId se presente", () => {
    render(<EventDetails event={{ ...baseEvent, proofId: "vc-555" }} />);
    expect(screen.getByText(/VC principale \(proofId\):/i)).toBeInTheDocument();
    expect(screen.getByText("vc-555")).toBeInTheDocument();
  });

  it("visualizza campo opzionale vcIds se presenti", () => {
    render(<EventDetails event={{ ...baseEvent, vcIds: ["vc-1", "vc-2"] }} />);
    expect(screen.getByText(/VC associate:/i)).toBeInTheDocument();
    expect(screen.getByText("vc-1, vc-2")).toBeInTheDocument();
  });

  it("renderizza null se event è null/undefined", () => {
    const { container } = render(<EventDetails event={undefined as any} />);
    expect(container.firstChild).toBeNull();
  });

  it("snapshot", () => {
    const { asFragment } = render(<EventDetails event={baseEvent} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
