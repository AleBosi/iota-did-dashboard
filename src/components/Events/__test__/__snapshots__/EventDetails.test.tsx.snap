// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EventDetails component snapshot 1`] = `
<DocumentFragment>
  <div
    class="border rounded p-4 bg-gray-50 mb-2"
  >
    <div>
      <b>
        ID evento:
      </b>
       evt-001
    </div>
    <div>
      <b>
        Tipo:
      </b>
       manutenzione
    </div>
    <div>
      <b>
        Descrizione:
      </b>
       Controllo annuale macchina
    </div>
    <div>
      <b>
        Prodotto:
      </b>
       prod-01
    </div>
    <div>
      <b>
        Operatore:
      </b>
       op-11
    </div>
    <div>
      <b>
        Macchinario:
      </b>
       mac-22
    </div>
    <div>
      <b>
        Creatore (DID):
      </b>
       did:creator-xyz
    </div>
    <div>
      <b>
        Data:
      </b>
       8/5/2025, 2:00:00 PM
    </div>
    <div>
      <b>
        Stato:
      </b>
       
      <span
        class="text-orange-700"
      >
        Da eseguire
      </span>
    </div>
  </div>
</DocumentFragment>
`;
